todocli/api_urls.py

```diff
@@ -0,0 +1,142 @@
+from enum import Enum
+
+base_api_url = "https://graph.microsoft.com/v1.0/me/todo"
+
+
+class ODataSystemQuery:
+    class Format(Enum):
+        json = "json"
+        atom = "atom"
+        xml = "xml"
+
+    class SortOrder(Enum):
+        Ascending = "asc"
+        Descending = "desc"
+
+    def add_to_query(self, expr):
+        if len(self.query_expr) == 1:
+            self.query_expr += expr
+        else:
+            self.query_expr += "&" + expr
+
+    def __init__(self):
+        self.query_expr = "?"
+
+    def count(self):
+        self.add_to_query("$count=true")
+        return self
+
+    def expand(self, expr):
+        self.add_to_query("$expand={}".format(expr))
+        return self
+
+    def filter_startsWith(self, property_name, expr):
+        self.filter("startsWith({},'{}')".format(property_name, expr))
+        return self
+
+    def filter_endsWith(self, property_name, expr):
+        self.filter("endsWith({},'{}')".format(property_name, expr))
+        return self
+
+    def _filter_compare(self, property_name, comparison, expr):
+        self.filter("{} {} '{}'".format(property_name, comparison, expr))
+        return self
+
+    def filter_eq(self, property_name, expr):
+        self._filter_compare(property_name, "eq", expr)
+        return self
+
+    def filter_ne(self, property_name, expr):
+        self._filter_compare(property_name, "ne", expr)
+        return self
+
+    def filter_gt(self, property_name, expr):
+        self._filter_compare(property_name, "gt", expr)
+        return self
+
+    def filter_ge(self, property_name, expr):
+        self._filter_compare(property_name, "ge", expr)
+        return self
+
+    def filter_lt(self, property_name, expr):
+        self._filter_compare(property_name, "lt", expr)
+        return self
+
+    def filter_le(self, property_name, expr):
+        self._filter_compare(property_name, "le", expr)
+        return self
+
+    def filter(self, expr):
+        self.add_to_query("$filter={}".format(expr))
+        return self
+
+    def format(self, ftmat: Format):
+        self.add_to_query("$format={}".format(ftmat.value))
+        return self
+
+    def orderBy(self, expr, sort_order: SortOrder = SortOrder.Ascending):
+        self.add_to_query("$orderby={} {}".format(expr, sort_order.value))
+        return self
+
+    def search(self, expr):
+        raise Exception("Not implemented")
+
+    def select(self, expr):
+        raise Exception("Not implemented")
+
+    def skip(self, expr):
+        raise Exception("Not implemented")
+
+    def top(self, num_elements):
+        self.add_to_query("$top={}".format(num_elements))
+        return self
+
+    def get(self):
```

Comment:
I've never worked with descriptor classes before, as far as I understand it, the __get__ function is only evaluated when assigning the descriptor class instance to a member of another class... (see https://stackoverflow.com/questions/54150798/python-get-is-not-called-for-my-class).

So if we changed it to a descriptor class, we'd need to assign the ODataSystemQuery instance to another classes member. But there is no other class we could assign the instance to in this context. 

We could assign it to a class member of ODataSystemQuery in "add_to_query" like this:

```    
def _add_to_query(self, expr):
        if len(self.query_expr) == 1:
            self.query_expr += expr
        else:
            self.query_expr += "&" + expr
        self.query_result = self
```

But then we still need to specify the property name "query_result" instead of ".get()" like this:
```
def query_task_by_name(todo_task_list_id, task_name: str):
    return (
        new_task(todo_task_list_id)
        + ODataSystemQuery().filter_eq("title", task_name).query_result
    )
```

Which is not really an advantage over calling ".get()".

Maybe I just lack the understanding on how to implement that properly... 
How would you do it?

Reply:
Hey I got confused - you're right, apologies. I was thinking of a cleaner way to do this either by:
- Let `ODataSystemQuery` inherit `str` or
- implement `__str__` in `ODataSystemQuery`
but I think both options make the code unnecessarily complex. Lets leave it at is, I think it's good. My bad!

Reply:
Okay, I'll just leave it as it is :+1: